tokio-trace/tokio-trace-core/src/subscriber.rs

```diff
@@ -262,6 +264,46 @@ pub trait Subscriber: 'static {
     fn drop_span(&self, id: span::Id) {
         let _ = id;
     }
+
+    // === Downcasting methods ================================================
+
+    /// If `self` is the same type as the provided `TypeId`, returns an untyped
+    /// `*const` pointer to that type. Otherwise, returns `None`.
+    ///
+    /// If you wish to downcast a `Subscriber`, it is strongly advised to use
+    /// the safe API provided by [`downcast_ref`] instead.
+    ///
+    /// This API is required for `downcast_raw` to be a trait method; a method
+    /// signature like [`downcast_ref`] (with a generic type parameter) is not
+    /// object-safe, and thus cannot be a trait method for `Subscriber`. This
+    /// means that if we only exposed `downcast_ref`, `Subscriber`
+    /// implementations could not override the downcasting behavior
+    ///
+    /// This method may be overridden by "fan out" or "chained" subscriber
+    /// implementations which consist of multiple composed types. Such
+    /// subscribers might allow `downcast_raw` by returning references to those
+    /// component if they contain components with the given `TypeId`.
+    fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
```

Comment:
I think this should be an `unsafe` function, because if it returns an invalid ptr, we enter UB. The `unsafe` block in `downcast_ref` assumes that the impl of `downcast_raw` is correct, so implementors should treat implementations as unsafe.

Reply:
Good point. Will make this unsafe.